### PR TITLE
Expose serial number and bus type via GuestOsInfo

### PR DESCRIPTION
**Pull request by alromeros**
**What this PR does / why we need it**:

This PR updates the guestOsInfo command to show additional disk info, like serial number and bus type.

Example:

```sh
# virtctl guestosinfo testvmi-7j99w
selecting docker as container runtime
{
  "guestAgentVersion": "6.1.0",
  "supportedCommands": [
 ...
  ],
  "hostname": "testvmi-7j99w",
  "os": {
    "name": "Fedora Linux",
    "kernelRelease": "5.14.10-300.fc35.x86_64",
    "version": "35 (Cloud Edition)",
    "prettyName": "Fedora Linux 35 (Cloud Edition)",
    "versionId": "35",
    "kernelVersion": "#1 SMP Thu Oct 7 20:48:44 UTC 2021",
    "machine": "x86_64",
    "id": "fedora"
  },
  "timezone": "UTC, 0",
  "fsInfo": {
    "disks": [
      {
        "diskName": "vda3",
        "mountPoint": "/boot/efi",
        "fileSystemType": "vfat",
        "usedBytes": 10227712,
        "totalBytes": 104607744,
        "disk": [
          {
            "serial": "WD-WMAP9A966149",
            "bus-type": "virtio"
          }
        ]
      },
...
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://issues.redhat.com/browse/CNV-18671

**Special notes for your reviewer**:

Added the api exceptions so we can fetch the guest-agent information without unnecessary structs as intermediaries (json names need to match). The api additions are in that way (Disk being an array, for example) to match the guest-agent response.

**Release note**:

```release-note
Expose fs disk information via GuestOsInfo
```

### CODE REVIEW

**Comment by enp0s3** [comment]
/hold

@mhenriks @alromeros Sorry perhaps I am missing something, since you've added a non-primitive type to the api, don't you miss the piece of code that should prepare the slice as part of the parsing process?

File: pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go [hunk 1]
```diff
@@ -196,6 +197,7 @@ func parseFilesystem(agentReply string) ([]api.Filesystem, error) {
 			Type:       fs.Type,
 			TotalBytes: fs.TotalBytes,
 			UsedBytes:  fs.UsedBytes,
+			Disk:       fs.Disk,
```
Comment (enp0s3) [inline pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:200]: Is it safe to just assign it like that w/o iterating and running append` to the list you create with a `make`?

Comment (alromeros) [reply pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:200]: Yeah I think it's safe to just assign slices

**Comment by alromeros** [comment]
> /hold
> 
> @mhenriks @alromeros Sorry perhaps I am missing something, since you've added a non-primitive type to the api, don't you miss the piece of code that should prepare the slice as part of the parsing process?

Hey @enp0s3, isn't that automatically handled by the generate scripts? What is exactly missing in this one?

**Comment by enp0s3** [comment]
@alromeros I took [GetUsers()](https://github.com/kubevirt/kubevirt/blob/0a3141fbb468107fa782b760fbc8076cb53d96f0/pkg/virt-launcher/virtwrap/manager.go#L2048) as a reference. The code iterates over the data it got from `l.agentData.GetUsers(-1)` , so how this is different from your case where you receive the `fs.disk` as a list?

Should you have a function `l.agentData.GetDisks(fs)` or something similar?

**Comment by enp0s3** [comment]
@alromeros I think I am confused because of the definition `[]v1.VirtualMachineInstanceFileSystemDisk`, why did you defined it as a slice? do you expect to receive a list of disks from the AgentData in `fs.Disk`?

**Comment by alromeros** [comment]
> @alromeros I took [GetUsers()](https://github.com/kubevirt/kubevirt/blob/0a3141fbb468107fa782b760fbc8076cb53d96f0/pkg/virt-launcher/virtwrap/manager.go#L2048) as a reference. The code iterates over the data it got from `l.agentData.GetUsers(-1)` , so how this is different from your case where you receive the `fs.disk` as a list?
> 
> Should you have a function `l.agentData.GetDisks(fs)` or something similar?

Right, I get what you mean. Following your example, `GetUsers` uses an [intermediary struct]( https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go#L195) on the agent-poller side to store the guest agent response. Then, it copies each element of the `api.User` type to the `v1.VirtualMachineInstanceGuestOSUser` type [here](https://github.com/kubevirt/kubevirt/blob/0a3141fbb468107fa782b760fbc8076cb53d96f0/pkg/virt-launcher/virtwrap/manager.go#L2049
).

We have something similar in [GetFilesystems](https://github.com/kubevirt/kubevirt/blob/0a3141fbb468107fa782b760fbc8076cb53d96f0/pkg/virt-launcher/virtwrap/manager.go#L2065C32-L2065C46) (the command I'm updating). We fetch the whole guest agent response [here](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go#L195). Since the new field I added (`disks`) has the same JSON formatting as the `disks` field in the guest agent response, there's no need for an intermediary struct and we can just assign it without needing to iterate.

**Comment by alromeros** [comment]
> @alromeros I think I am confused because of the definition `[]v1.VirtualMachineInstanceFileSystemDisk`, why did you defined it as a slice? do you expect to receive a list of disks from the AgentData in `fs.Disk`?

Right, I'm just mirroring the guest agent's response (an array of `disk`).

**Comment by enp0s3** [comment]
/unhold

@alromeros Thank you for the explanation! So the `json.Unmarshal([]byte(response), &result)` already allocates the slice.

File: staging/src/kubevirt.io/api/core/v1/types.go [hunk 2]
```diff
@@ -2215,13 +2215,20 @@ type VirtualMachineInstanceFileSystemList struct {
 	Items           []VirtualMachineInstanceFileSystem `json:"items"`
 }
 
+// VirtualMachineInstanceFileSystemDisk represents the guest os FS disks
+type VirtualMachineInstanceFileSystemDisk struct {
+	Serial  string `json:"serial"`
```
Comment (xpivarc) [inline staging/src/kubevirt.io/api/core/v1/types.go:2220]: Missing `omitempty`

File: staging/src/kubevirt.io/api/core/v1/types.go [hunk 3]
```diff
@@ -2215,13 +2215,20 @@ type VirtualMachineInstanceFileSystemList struct {
...
+	BusType string `json:"bus-type"`
```
Comment (xpivarc) [inline staging/src/kubevirt.io/api/core/v1/types.go:2221]: This does break convention, it should be `json:"busType"`.

File: staging/src/kubevirt.io/api/core/v1/types.go [hunk 4]
```diff
@@ -2215,13 +2215,20 @@ type VirtualMachineInstanceFileSystemList struct {
...
-	MountPoint     string `json:"mountPoint"`
-	FileSystemType string `json:"fileSystemType"`
-	UsedBytes      int    `json:"usedBytes"`
-	TotalBytes     int    `json:"totalBytes"`
+	DiskName       string                                 `json:"diskName"`
+	MountPoint     string                                 `json:"mountPoint"`
+	FileSystemType string                                 `json:"fileSystemType"`
+	UsedBytes      int                                    `json:"usedBytes"`
+	TotalBytes     int                                    `json:"totalBytes"`
+	Disk           []VirtualMachineInstanceFileSystemDisk `json:"disk,omitempty"`
```
Comment (xpivarc) [inline staging/src/kubevirt.io/api/core/v1/types.go:2231]: Why is this a slice?

**Review by xpivarc — COMMENTED** [review]
@enp0s3 this PR doesn't add a new test nor does it extend the existing functional test.

File: staging/src/kubevirt.io/api/core/v1/types.go (hunk 3 above)
Comment (enp0s3) [reply staging/src/kubevirt.io/api/core/v1/types.go:2221]: @xpivarc Thanks for the heads up! this check must be automated somehow, it is something hard to remember for the long term.

**Comment by enp0s3** [comment]
> @enp0s3 this PR doesn't add a new test nor does it extend the existing functional test.

@xpivarc TBH I felt uncomfortable with that as well but I had a doubt whether functional coverage fits here, or unit is enough.

File: staging/src/kubevirt.io/api/core/v1/types.go (hunk 4 above)
Comment (enp0s3) [reply staging/src/kubevirt.io/api/core/v1/types.go:2231]: I got the explanation [here](https://github.com/kubevirt/kubevirt/pull/10970#issuecomment-1905585140)

Comment (xpivarc) [reply staging/src/kubevirt.io/api/core/v1/types.go:2231]: That doesn't explain the question, @alromeros can you elaborate?

File: staging/src/kubevirt.io/api/core/v1/types.go (hunk 3 above)
Comment (xpivarc) [reply staging/src/kubevirt.io/api/core/v1/types.go:2221]: we can make the `make generate` fail

**Comment by xpivarc** [comment]
> > @enp0s3 this PR doesn't add a new test nor does it extend the existing functional test.
> 
> @xpivarc TBH I felt uncomfortable with that as well but I had a doubt whether functional coverage fits here, or unit is enough.

@enp0s3 
There is an existing functional test, all that is needed to include check that the proper field is populated. Zero cost.

File: staging/src/kubevirt.io/api/core/v1/types.go (hunk 4 above)
Comment (alromeros) [reply staging/src/kubevirt.io/api/core/v1/types.go:2231]: We are storing here the guest agent response to the `guest-get-fsinfo` command. The command returns an array of disks, so I'm using one: https://qemu-project.gitlab.io/qemu/interop/qemu-ga-ref.html#qapidoc-161.

**Comment by alromeros** [comment]
> > > @enp0s3 this PR doesn't add a new test nor does it extend the existing functional test.
> > 
> > 
> > @xpivarc TBH I felt uncomfortable with that as well but I had a doubt whether functional coverage fits here, or unit is enough.
> 
> @enp0s3 There is an existing functional test, all that is needed to include check that the proper field is populated. Zero cost.

Didn't see individual checks for most of the fields in fsList but I can add them in a follow-up. Will open it later once we decide if the API is good.

File: staging/src/kubevirt.io/api/core/v1/types.go (hunk 2 above)
Comment (mhenriks) [reply staging/src/kubevirt.io/api/core/v1/types.go:2220]: @alromeros is this ever expected to be empty?  I would be pretty surprised

File: staging/src/kubevirt.io/api/core/v1/types.go (hunk 3 above)
Comment (mhenriks) [reply staging/src/kubevirt.io/api/core/v1/types.go:2221]: I assumed we wanted to expose with exact naming naming scheme as libvirt

Comment (alromeros) [reply staging/src/kubevirt.io/api/core/v1/types.go:2221]: Yeah me too. It prevents adding some unnecessary code, but I guess it's odd to only add the exception to BusType.

File: staging/src/kubevirt.io/api/core/v1/types.go (hunk 2 above)
Comment (alromeros) [reply staging/src/kubevirt.io/api/core/v1/types.go:2220]: I think BusType will never be empty, but Serial is optional: https://qemu-project.gitlab.io/qemu/interop/qemu-ga-ref.html#qapidoc-146.

**Pull request merged**